mlir/include/mlir/IR/PatternMatch.h

```diff
@@ -1457,6 +1457,33 @@ class PDLPatternModule {
                                    std::forward<ConstraintFnT>(constraintFn)));
   }
 
+  /// Register a constraint function that produces results with PDL. A
+  /// constraint function with results uses the same type and registry as
+  /// rewrite functions. It may be specified in one of two ways:
+  ///
+  ///   * `void (PatternRewriter &, PDLResultList &, ArrayRef<PDLValue>)`
+  ///
+  ///   In this overload the arguments of the constraint function are passed via
+  ///   the low-level PDLValue form, and the results are manually appended to
+  ///   the given result list.
+  ///
+  ///   * `ResultT (PatternRewriter &, ValueTs... values)`
```

Comment:
How do we signal a `failure` with this interface? Don't we need something similar like `FailureOr<ResultT>`?

Reply:
Good catch!
I think it makes sense to remove the template magic and under the hood conversion that is enabled by this. The problem is that not all valid native rewrite functions are also valid constraint with results functions. We would have to adjust how the conversion from PDL structures to built-in structure is done for this.
So for the scope of this PR we will only support constraints with results in form of
`LogicalResult (PatternRewriter &, PDLResultList &, ArrayRef<PDLValue>)`